### PR TITLE
CB-20947 Additional metrics for periscope to Prometheus

### PR DESCRIPTION
**Pull request by aakashshankar**
See detailed description in the commit message.

### CODE REVIEW

**Comment by aakashshankar** [comment]
retest this please

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java [hunk 1]
```diff
@@ -132,6 +140,11 @@ protected void execute() {
             }
         } catch (Exception ex) {
             LOGGER.info("Failed to process load alert for Cluster '{}', userCrn: {}, exception: ", stackCrn, pollingUserCrn, ex);
+            if (ex instanceof ForbiddenException) {
+                metricService.incrementMetricCounter(YARN_FORBIDDEN_EXCEPTION, MetricTag.TENANT.name(),
+                        Optional.ofNullable(cluster.getClusterPertain().getTenant()).orElse("NA"), MetricTag.CLOUD_PROVIDER.name(),
+                        Optional.ofNullable(cluster.getCloudPlatform()).orElse("NA"));
+            }
```
Comment (adarshsharma1983) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java:147]: shouldn't we have this in YarnMetricsClient since there is actual call happening. Here we are assuming that ForbiddenException will be propagated which can change in future. Also, metric should be generated at the same place where actual operation is being executed.

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java [hunk 2]
```diff
@@ -33,17 +42,26 @@ public FreeIpaCommunicator(UserV1Endpoint userV1Endpoint) {
...
+                        internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
+                        () -> userV1Endpoint.synchronizeAllUsers(request));
+                metricService.incrementMetricCounter(IPA_USER_SYNC_INVOCATION, MetricTag.TENANT.name(), request.getAccountId());
+                return status;
+            }
+        } catch (Exception ex) {
+            LOGGER.error("Encountered error when trying to synchronize users to IPA for environment: {}", envCrn, ex);
+            metricService.incrementMetricCounter(IPA_USER_SYNC_FAILED, MetricTag.TENANT.name(), request.getAccountId());
+
+            throw new RuntimeException(format("Failed to perform user sync for environment: %s", envCrn));
```
Comment (adarshsharma1983) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java:64]: metric will also be incremented even if getLastSyncOperationStatus call fails so it will be treated user sync failed which was actually never attempted

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ClusterService.java [hunk 3]
```diff
@@ -318,4 +323,11 @@ private void calculateClusterStateMetrics() {
         periscopeMetricService.submit(MetricType.CLUSTER_STATE_SUSPENDED,
                 clusterRepository.countByStateAndAutoscalingEnabledAndPeriscopeNodeId(SUSPENDED, true, periscopeNodeConfig.getId()));
     }
+
+    private void calculateUnprocessedScalingActivityMetrics() {
+        // TODO: Should we fetch a broader collection of scaling events?
+        Collection<Long> activityIds = scalingActivityService.findAllInStatusesThatStartedBefore(ActivityStatus.getStatusesApplicableForCleanup(),
+                UNPROCESSED_SCALING_ACTIVITY_AGE_HOURS);
+        periscopeMetricService.submit(MetricType.STALE_SCALING_ACTIVITY, activityIds.size());
```
Comment (adarshsharma1983) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ClusterService.java:331]: This metric will be generated only once as part of each deployment/restart. Can you elaborate the reasoning for it?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java [hunk 4]
```diff
@@ -32,4 +39,26 @@ protected void init() {
     protected Optional<String> getMetricPrefix() {
         return Optional.of(METRIC_PREFIX);
     }
+
+    @Override
+    protected boolean gaugeMetric(Metric metric) {
+        return metric.getMetricName().contains("state") || metric.getMetricName().contains("leader") || metric.getMetricName().contains("threadpool")
+                || metric.getMetricName().contains("scaling.activity");
```
Comment (adarshsharma1983) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java:46]: Can we define a constant will all these values and update this code. In future, only value for constant will change.

File: autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java [hunk 5]
```diff
@@ -32,4 +39,26 @@ protected void init() {
...
+    }
+
+    private String[] createTags(Cluster cluster) {
+        return new String[]{MetricTag.TENANT.name(), Optional.ofNullable(cluster.getClusterPertain().getTenant()).orElse("NA"),
```
Comment (adarshsharma1983) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java:50]: Can cluster.getClusterPertain() be null?

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java [hunk 6]
```diff
@@ -33,17 +42,26 @@ public FreeIpaCommunicator(UserV1Endpoint userV1Endpoint) {
...
+            } else {
+                SyncOperationStatus status = ThreadBasedUserCrnProvider.doAsInternalActor(
+                        internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
+                        () -> userV1Endpoint.synchronizeAllUsers(request));
+                metricService.incrementMetricCounter(IPA_USER_SYNC_INVOCATION, MetricTag.TENANT.name(), request.getAccountId());
+                return status;
+            }
+        } catch (Exception ex) {
+            LOGGER.error("Encountered error when trying to synchronize users to IPA for environment: {}", envCrn, ex);
+            metricService.incrementMetricCounter(IPA_USER_SYNC_FAILED, MetricTag.TENANT.name(), request.getAccountId());
```
Comment (Vatsal23082000) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java:62]: If the SynchronizationStatus is REJECTED than should it be considered in the failed one as in this case the sync process might not have been performed.

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ClusterService.java (hunk 3 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/ClusterService.java]: This should be updated periodically as done in `ClusterSerivce`. Refactored this into `ScalingActivityService` instead, following the same pattern

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java (hunk 2 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java]: Refactored to detect only if IPA user sync failed

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java (hunk 1 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java:147]: There are CM and Yarn API calls being made in the corresponding try block, along with a CB API call to retrieve stackv4response. We want to process all possible exceptions thrown from those invocations to this catch block and publish an UpdateFailedEvent. Modifying this logic to detect the exception in YarnMetricsClient would require another exception to be re-thrown which will eventually be caught here. Additionally, the handler for UpdateFailedEvent (UpdateFailedHandler) has some logic that processes these Forbidden Exceptions.

File: autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java (hunk 5 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java:50]: We had seen issues previously where the ClusterPertain entity did not have correct unique constraints when being persisted into the DB. I added a null check to be safe. Also, I followed the pattern on adding such tags for metrics from the core service

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java (hunk 6 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java:72]: That is already being done because it looks like the REJECTED state is set if a SyncOperationAlreadyRunningException is thrown. The exception is then handled to increment failure count.

https://github.com/hortonworks/cloudbreak/blob/25f11bef48b7b9b2cb7985da70f6538b734b122a/freeipa/src/main/java/com/sequenceiq/freeipa/controller/UserV1Controller.java#L202-L207

**Comment by aakashshankar** [comment]
Tests are failing because repo.hortonworks.com is down

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java (hunk 1 above)
Comment (sidseth) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java:147]: This should be in the YarnMetricsClient. Also separately for the CM invocations. The bit about the UpdateFailedHandler tracking a single error is something we should address in a separate jira. (The invocation of YARN vs CM uses a different user; it doesn't make sense to change this user if the CM API throws a 403).

Additionally, we want to track 1) Overall errors from these API invocations, and 2) sub-categorize the 403 errors since those are the most problematic right now. (Not sure if we can extract the error code and use that as a tag value, or we need to have a separate metric/tag based on the ForbiddenException)

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java [hunk 7]
```diff
@@ -33,17 +42,36 @@ public FreeIpaCommunicator(UserV1Endpoint userV1Endpoint) {
...
+                    internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
+                    () -> userV1Endpoint.getLastSyncOperationStatus(envCrn));
+            if (SynchronizationStatus.RUNNING.equals(lastSyncStatus.getStatus())) {
+                LOGGER.info("There is a user sync operation already running for environment: {} with operationId: {}, " +
+                        "skipping request to trigger another user sync", envCrn, lastSyncStatus.getOperationId());
+                return lastSyncStatus;
+            } else {
+                return invokeFreeipaUserSyncAndHandleException(request, envCrn);
+            }
+        } catch (Exception ex) {
```
Comment (sidseth) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java:56]: Why is this catch block required here?
As things are set up right now - if there's an error from userV1Endpoint.synchronizeAllUsers - it is going to end up being wrapped in 2 RuntimExceptions.

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java [hunk 8]
```diff
@@ -33,17 +42,36 @@ public FreeIpaCommunicator(UserV1Endpoint userV1Endpoint) {
...
+            SyncOperationStatus status = ThreadBasedUserCrnProvider.doAsInternalActor(
                     internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
                     () -> userV1Endpoint.synchronizeAllUsers(request));
+            metricService.incrementMetricCounter(IPA_USER_SYNC_INVOCATION, MetricTag.TENANT.name(), request.getAccountId());
+            return status;
+        } catch (Exception ex) {
+            LOGGER.error("Failed to synchronize users to IPA for environment: {}", envCrn, ex);
+            metricService.incrementMetricCounter(IPA_USER_SYNC_FAILED, MetricTag.TENANT.name(), request.getAccountId());
+
+            throw new RuntimeException(format("User sync failed for environment %s", envCrn), ex);
```
Comment (sidseth) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java:74]: The invoked method is not set to throw checked exceptions. To avoid unnecessary nesting of Exception (and propagating the original Exception) - this should check if the Exception is a RuntimeException - if it is, cast and rethrow, if not - then convert it to an explicit RuntimeException.

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java (hunk 6 above)
Comment (sidseth) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java:72]: This looks ok to me as is. The API itself is not very well defined in terms of what SyncStatus can be returned. Beyond REJECTED, there's other states like FAILED. Do we need to be checking for these? (The API is not very clear on this)

File: autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java (hunk 5 above)
Comment (sidseth) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java:50]: Are we adding the null check on it? Can be another one which goes into the "NA" category ?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java [hunk 9]
```diff
@@ -190,14 +205,22 @@ public Set<Long> findAllInStatusesThatStartedBefore(Collection<ActivityStatus> s
...
     }
 
-    public void save(ScalingActivity scalingActivity) {
-        scalingActivityRepository.save(scalingActivity);
+    public ScalingActivity save(ScalingActivity scalingActivity) {
+        return scalingActivityRepository.save(scalingActivity);
+    }
+
+    private void calculateUnprocessedScalingActivityMetrics() {
+        // TODO: Should we fetch a broader collection of scaling events?
```
Comment (sidseth) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java:221]: This, I think, is the opposite of what we want to track. The events which are applicable for cleanup - will be cleaned up (assuming the system is not too busy).

What would be useful to track is: 
1) How many operations exist which are waiting for an activity to complete (i.e. are not cleanup candidates) - and effectively cause the table to grow continuously.
2) The total size of this table
3) Maybe this metric: the total number of cleanup candidates.

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java [hunk 10]
```diff
@@ -190,14 +205,22 @@ public Set<Long> findAllInStatusesThatStartedBefore(Collection<ActivityStatus> s
...
+        Collection<Long> activityIds = findAllInStatusesThatStartedBefore(ActivityStatus.getStatusesApplicableForCleanup(),
```
Comment (sidseth) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java:222]: Getting this as a Collection, and then counting it is not great. This needs to be a count at the SQL level itself.

Which brings up another aspect. The CleanupMonitor is going to perform the exact same query. So if, for some reason, the cleanup monitor has not run in some time - it could end up pulling in enough events to overwhelm the system. @aakashshankar  - can you please file a separate jira to track this, where the CleanupMonitor limits the number of events it tries to clean up in a single run ?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java [hunk 11]
```diff
@@ -103,6 +117,7 @@ public void update(Long activityId, FlowIdentifier flowIdentifier, ActivityStatu
         scalingActivity.setActivityStatus(activityStatus);
         scalingActivity.setScalingActivityReason(reason);
         LOGGER.info("Updating ScalingActivity: {} with FlowInformation and ActivityStatus: {}", activityId, activityStatus);
+        calculateUnprocessedScalingActivityMetrics();
```
Comment (sidseth) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java:120]: Performing this operation on every DB operation - would this get expensive ?
Instead could we have these metrics piggyback on one of the existing monitors, which runs more frequently - or even set up a separate Monitor / Timed Executor which updates these metrics ?

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/client/YarnMetricsClient.java [hunk 12]
```diff
@@ -65,7 +70,10 @@ public YarnScalingServiceV1Response getYarnMetricsForCluster(Cluster cluster, St
                 tlsConfig.getClientCert(), tlsConfig.getClientKey(), true);
         String yarnApiUrl = String.format(YARN_API_URL, clusterProxyUrl, cluster.getStackCrn());
 
+        long start = Instant.now().toEpochMilli();
```
Comment (sidseth) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/client/YarnMetricsClient.java:73]: Use Clock here as well - similar to what is done in YarnBasedScalingAdjustmentService ?

File: autoscale/src/test/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentServiceTest.java [hunk 13]
```diff
@@ -433,6 +437,7 @@ private void setupBasicMocks(StackV4Response stackResponse, YarnScalingServiceV1
         doReturn(yarnResponse).when(yarnMetricsClient).getYarnMetricsForCluster(any(Cluster.class), eq(stackResponse), eq(HOSTGROUP), eq(POLLING_USER_CRN),
                 any(Optional.class));
         doReturn(TEST_MESSAGE).when(messagesService).getMessageWithArgs(anyString(), any());
+        doNothing().when(metricService).recordYarnInvocation(any(Cluster.class), anyLong());
```
Comment (sidseth) [inline autoscale/src/test/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentServiceTest.java:440]: Can we verify some of the expected metrics being populated ?

**Review by sidseth — COMMENTED** [review]
From the jira, a few additional metrics need to be tracked (or maybe these are already tracked).
* Scaling operations in the past X minutes, by type
* Time spent within Periscope for a scaling trigger.
* Number of clusters with different types of autoscaling configured

Some of these could be via a follow up jira.

Additionally, would be useful to track
* Whether the cleanupHandler thread was invoked (last invocation time or total invocation count) - this thread dying will lead to eventual problems.

**Comment by sidseth** [comment]
Also, for the 403 error tracking - we just noticed that there's occasional invocations where the initial request receives a 403. The overall operation eventually succeeds because of retries. Need to be tracking these individual invocations as well - especially since there's no good reason for these 403s.
This, I think, will be far more involved - and

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java (hunk 1 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java]: Moved this to YarnMetricsClient with exception handling

File: autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java (hunk 4 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java]: Updated

File: autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java (hunk 5 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/PeriscopeMetricService.java]: Added an explicit null check

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java (hunk 7 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java]: Changed this

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java (hunk 9 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java]: Tracked all these metrics

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java (hunk 10 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java]: Modified to count at the SQL level, also filed https://jira.cloudera.com/browse/CB-21498 to limit the events cleaned up by CleanupMonitor

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java (hunk 11 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java]: Moved this to CleanupMonitor to run every 10 mins. I initially added this here because a similar pattern is followed in `ClusterService`, where metrics are calculated for every DB operation

File: autoscale/src/test/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentServiceTest.java (hunk 13 above)
Comment (aakashshankar) [reply autoscale/src/test/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentServiceTest.java:446]: Added verification

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java [hunk 14]
```diff
@@ -57,6 +57,9 @@ public class ScalingActivityService implements AuthorizationResourceCrnProvider,
     @Inject
     private ClusterService clusterService;
 
+    @Inject
```
Comment (Vatsal23082000) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java:60]: Can we remove this inject as it is not being used here?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java [hunk 15]
```diff
@@ -197,7 +208,7 @@ public void deleteScalingActivityForCluster(Long clusterId) {
         scalingActivityRepository.deleteAllByCluster(clusterId);
     }
 
-    public void save(ScalingActivity scalingActivity) {
-        scalingActivityRepository.save(scalingActivity);
+    public ScalingActivity save(ScalingActivity scalingActivity) {
```
Comment (Vatsal23082000) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java:211]: why did you change this function to return a scaling activity?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java [hunk 16]
```diff
@@ -177,6 +180,14 @@ public List<ScalingActivity> findAllByIds(Collection<Long> ids) {
         return newArrayList(scalingActivityRepository.findAllById(ids));
     }
 
+    public Long countAllActivitiesInStatuses(Collection<ActivityStatus> statuses) {
```
Comment (Vatsal23082000) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java:183]: Can this be covered under unit tests?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java (hunk 14 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java]: 👍🏽

**Review by Vatsal23082000 — APPROVED** [review]
LGTM

**Review by adarshsharma1983 — APPROVED** [review]

**Review by lajosrodek — APPROVED** [review]
Nice work, LGTM
